### PR TITLE
AZP: Build via dockers

### PR DESCRIPTION
**Pull request by avildema**
build UCX in dockers on AZP

### CODE REVIEW

File: buildlib/azure-pipelines-pr.yml [hunk 1]
```diff
@@ -76,6 +79,25 @@ stages:
...
+              CONTAINER: centos7_cuda11_1
+              CONFIGURE_OPTS:
+        container: $[ variables['CONTAINER'] ]
+
+        steps:
+          - checkout: self
+            clean: true
+
+          - bash: ./autogen.sh
+            displayName: Setup autotools
```
Comment (yosefe) [inline buildlib/azure-pipelines-pr.yml:99]: this step can be bigger, maybe use additional yml file?

File: buildlib/dockers/centos8-release.Dockerfile [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+    tar -xzf ${MOFED_IMAGE} && \
+    ${MOFED_DIR}/mlnxofedinstall --all -q \
+        --user-space-only \
+        --without-fw-update \
+        --skip-distro-check \
+        --without-ucx \
+        --without-hcoll \
+        --without-openmpi \
+        --without-sharp \
+    && rm -rf ${MOFED_DIR} && rm -rf *.tgz
```
Comment (yosefe) [inline buildlib/dockers/centos8-release.Dockerfile:43]: can we avoid code duplication? seems every distro is same way to install MOFED

File: buildlib/dockers/docker-compose.yml [hunk 3]
```diff
@@ -0,0 +1,89 @@
...
+  #   image: sles12-mofed5.1-cuda11.1
+  #   build:
+  #     context: .
+  #     network: host
+  #     dockerfile: sles-release.Dockerfile
+  #     args:
+  #       MOFED_VERSION: 5.1-2.5.8.0
+  #       OS_VERSION: 15.2
+  #       CUDA_VERSION: 11.1
+  #       MOFED_OS: sles15sp2
```
Comment (yosefe) [inline buildlib/dockers/docker-compose.yml:89]: remove?

File: buildlib/azure-pipelines-pr.yml (hunk 1 above)
Comment (avildema) [reply buildlib/azure-pipelines-pr.yml]: it just for debug, will launch like in jenkins via test_jenkins.sh (it've done)

File: buildlib/dockers/centos8-release.Dockerfile (hunk 2 above)
Comment (avildema) [reply buildlib/dockers/centos8-release.Dockerfile:43]: As I know dockerfile don't have support include blocks

File: buildlib/dockers/docker-compose.yml (hunk 3 above)
Comment (avildema) [reply buildlib/dockers/docker-compose.yml:89]: It's question, sles don't have official docker container

**Comment by avildema** [comment]
bot:pipe:retest

File: buildlib/azure-pipelines-pr.yml [hunk 4]
```diff
@@ -17,6 +17,27 @@ resources:
     - container: coverity_rh7
       image: rdmz-harbor.rdmz.labs.mlnx/ucx/coverity:mofed-5.1-2.3.8.0
       options: -v /hpc/local:/hpc/local -v /auto/sw_tools:/auto/sw_tools
+    - container: rhel76
+      image: rdmz-harbor.rdmz.labs.mlnx/swx-infra/x86_64/rhel7.6/builder:mofed-5.1-2.3.8.0
```
Comment (yosefe) [inline buildlib/azure-pipelines-pr.yml:21]: need to build with mofed 5.0 so it will be compatible with mofed 5.x

File: buildlib/tools/builds.sh [hunk 5]
```diff
@@ -0,0 +1,428 @@
...
+build_java_docs() {
+	echo " ==== Building java docs ===="
+	if az_module_load dev/jdk && az_module_load dev/mvn
+	then
+		../configure --prefix=$ucx_inst --with-java
+		$MAKE -C ../build-test/bindings/java/src/main/native docs
+		module unload dev/jdk
+		module unload dev/mvn
+	fi
+}
```
Comment (yosefe) [inline buildlib/tools/builds.sh:48]: should be part of jucx job
@petro-rudenko is it done already there?

File: buildlib/tools/builds.sh [hunk 6]
```diff
@@ -0,0 +1,428 @@
...
+
+#
+# Build with PGI compiler
+#
+build_pgi() {
+	echo 1..1 > build_pgi.tap
+	pgi_test_file=$(mktemp ./XXXXXX).c
+	echo "int main() {return 0;}" > ${pgi_test_file}
+
+	if az_module_load pgi/latest && pgcc18 --version && pgcc18 ${pgi_test_file} -o ${pgi_test_file}.out
```
Comment (yosefe) [inline buildlib/tools/builds.sh:155]: load specific version

File: buildlib/tools/builds.sh [hunk 7]
```diff
@@ -0,0 +1,428 @@
...
+	#see https://stackoverflow.com/questions/9705660/check-glibc-version-for-a-particular-gcc-compiler
+	if [ `cat /etc/os-release | grep -i "ubuntu\|mint"|wc -l` -gt 0 ]; then
+		azure_log_warning "==== Not building with latest gcc compiler ===="
+		echo "ok 1 - # SKIP because dev/gcc-latest module support only RHEL" >> build_gcc_latest.tap
+		return 0
+	fi
+	ldd_ver="$(ldd --version | awk '/ldd/{print $NF}')"
+	if (echo "2.14"; echo $ldd_ver) | sort -CV
+	then
+		if az_module_load dev/gcc-latest
```
Comment (yosefe) [inline buildlib/tools/builds.sh:305]: load specific version

File: buildlib/tools/builds.sh [hunk 8]
```diff
@@ -0,0 +1,428 @@
...
+}
+
+#
+# Build with armclang compiler
+#
+build_armclang() {
+	echo 1..1 > build_armclang.tap
+	armclang_test_file=$(mktemp ./XXXXXX).c
+	echo "int main() {return 0;}" > ${armclang_test_file}
+	if az_module_load arm-compiler/latest && armclang --version && armclang ${armclang_test_file} -o ${armclang_test_file}.out
```
Comment (yosefe) [inline buildlib/tools/builds.sh:333]: load specific version

File: buildlib/tools/builds.sh [hunk 9]
```diff
@@ -0,0 +1,428 @@
...
+		echo "==== Build with armclang compiler ===="
+		../contrib/configure-devel --prefix=$ucx_inst CC=armclang CXX=armclang++
+		make_clean
+		$MAKEP
+		$MAKEP install
+		make_clean distclean
+		echo "ok 1 - build successful " >> build_armclang.tap
+	else
+		echo "==== Not building with armclang compiler ===="
+		echo "ok 1 - # SKIP because armclang not installed" >> build_armclang.tap
```
Comment (yosefe) [inline buildlib/tools/builds.sh:344]: can probably remove all .tap stuff - it was for Jenkins

File: buildlib/tools/builds.sh [hunk 10]
```diff
@@ -0,0 +1,428 @@
...
+	fi
+
+	rm -rf ${armclang_test_file} ${armclang_test_file}.out
+	az_module_unload arm-compiler/latest
+}
+
```
Comment (yosefe) [inline buildlib/tools/builds.sh:350]: remove one space line

File: buildlib/tools/builds.sh [hunk 11]
```diff
@@ -0,0 +1,428 @@
...
+	else
+		log_error "Missing include config.h in files: $missing"
+		exit 1
+	fi
+}
+
+#
+# Do a given task only if the current worker is supposed to do it.
+#
+do_distributed_task() {
```
Comment (yosefe) [inline buildlib/tools/builds.sh:387]: do_task

File: buildlib/tools/builds.sh (hunk 5 above)
Comment (petro-rudenko) [reply buildlib/tools/builds.sh:48]: Yes, it's a part of AZP

Comment (yosefe) [reply buildlib/tools/builds.sh:48]: @avildema so let's remove it from here

File: buildlib/tools/common.sh [hunk 12]
```diff
@@ -7,6 +7,10 @@ GDRCOPY_MODULE="dev/gdrcopy-latest"
 JDK_MODULE="dev/jdk"
 MVN_MODULE="dev/mvn"
 XPMEM_MODULE="dev/xpmem-latest"
+PGI_MODULE="pgi/19.7"
+GCC_MODULE="dev/gcc-10.1.0"
+ARM_MODULE="arm-compiler/armcc-19.0"
+INTEL_MODULE="intel/ics-19.1.1"
```
Comment (yosefe) [inline buildlib/tools/common.sh:13]: nice!

File: buildlib/azure-pipelines-pr.yml (hunk 4 above)
Comment (yosefe) [reply buildlib/azure-pipelines-pr.yml:20]: unresolved

File: buildlib/tools/builds.sh (hunk 10 above)
Comment (yosefe) [reply buildlib/tools/builds.sh:307]: unresolved

**Review by yosefe — COMMENTED** [review]
- mofed 5.0
- cuda 11.2

**Comment by avildema** [comment]
> * mofed 5.0
> * cuda 11.2

@yosefe for build we can use cuda from module

**Comment by yosefe** [comment]
> > * mofed 5.0
> > * cuda 11.2
> 
> @yosefe for build we can use cuda from module

ok

**Review by yosefe — APPROVED** [review]
pls squash

**Comment by avildema** [comment]
bot:pipe:retest

**Comment by avildema** [comment]
bot:pipe:retest

**Comment by yosefe** [comment]
@avildema in next PR let's try to remove these:
image

**Review by yosefe — APPROVED** [review]

**Pull request merged**